libraries/cyclestreets-view/src/main/java/net/cyclestreets/views/overlay/WaymarkOverlay.kt

```diff
@@ -0,0 +1,159 @@
+import net.cyclestreets.view.R
+import net.cyclestreets.views.CycleMapView
+
+import org.osmdroid.api.IGeoPoint
+import org.osmdroid.api.IProjection
+import org.osmdroid.util.GeoPoint
+import org.osmdroid.views.MapView
+import org.osmdroid.views.overlay.Overlay
+import org.osmdroid.views.overlay.OverlayItem
+
+import java.util.ArrayList
+
+class WaymarkOverlay(private val mapView: CycleMapView) : Overlay(), PauseResumeListener, Route.Listener {
+
+    private val greenWisp: Drawable?
+    private val orangeWisp: Drawable?
+    private val redWisp: Drawable?
+    private val screenPos = Point()
+    private val bitmapTransform = Matrix()
+    private val bitmapPaint = Paint()
+
+    private val waymarkers: MutableList<OverlayItem>
+
+    init {
+        val res = mapView.context.resources
+        val makeWisp = { drawable: Int -> ResourcesCompat.getDrawable(res, drawable, null) }
+
+        greenWisp = makeWisp(R.drawable.green_wisp)
+        orangeWisp = makeWisp(R.drawable.orange_wisp)
+        redWisp = makeWisp(R.drawable.red_wisp)
+
+        waymarkers = ArrayList()
+    }
+
+    fun waymarkersCount(): Int {
+        return waymarkers.size
+    }
+
+    fun waypoints(): Waypoints {
+        return Waypoints(waymarkers.map { wp -> wp.point })
+    }
+
+    fun finish(): IGeoPoint {
+        return waymarkers.last().point
+    }
+
+    fun addWaypoint(point: IGeoPoint?) {
+        if (point == null)
+            return
+        when (waymarkersCount()) {
+            0 -> pushMarker(point, "start", greenWisp)
+            1 -> pushMarker(point, "finish", redWisp)
+            else -> {
+                val prevFinished = finish()
+                popMarker()
+                pushMarker(prevFinished, "waypoint", orangeWisp)
+                pushMarker(point, "finish", redWisp)
+            }
+        }
+    }
+
+    fun removeWaypoint() {
+        when (waymarkersCount()) {
+            0 -> { }
+            1, 2 -> popMarker()
+            else -> {
+                popMarker()
+                val prevFinished = finish()
+                popMarker()
+                pushMarker(prevFinished, "finish", redWisp)
+            }
+        }
+    }
+
+    private fun pushMarker(point: IGeoPoint, label: String, icon: Drawable?) {
+        waymarkers.add(makeMarker(point, label, icon))
+    }
+
+    private fun popMarker() {
+        waymarkers.removeAt(waymarkers.lastIndex)
+    }
+
+    private fun makeMarker(point: IGeoPoint, label: String, icon: Drawable?): OverlayItem {
+        return OverlayItem(label, label, GeoPoint(point.latitude, point.longitude)).apply {
+            setMarker(icon)
+            markerHotspot = OverlayItem.HotspotPlace.BOTTOM_CENTER
+        }
+    }
+
+    ////////////////////////////////////////////
+    override fun draw(canvas: Canvas, mapView: MapView, shadow: Boolean) {
+        val projection = mapView.projection
+
+        waymarkers.forEach { wp -> drawMarker(canvas, projection, wp) }
```

Comment:
Maybe inline `mapView.projection` in here?

Reply:
mapView.projection returns a new object on each invocation, and object allocation when drawing is generally considered a bad thing. (Or at least was - whether it's still an issue on modern phones who knows?)  Hoisting the call out of the forEach ensures we only call it once.

Reply:
Ah OK, I didn't know that!